### PR TITLE
Mapzen services use quotas

### PR DESCRIPTION
**Pull request by ethervoid**
- Now the mapzen services uses the user quota instead the hard limit we have in cdb_conf
- Refactored and added new tests
- Now routing config uses soft_limit for Mapzen provider

### CODE REVIEW

**Comment by jgoizueta** [comment]
LGTM :+1:
By the way, nice cleaning up of service/provider confusion in the tests...

**Comment by ethervoid** [comment]
Thank you for the review and for the comment 😃

File: server/lib/python/cartodb_services/cartodb_services/refactor/service/mapzen_geocoder_config.py [hunk 1]
```diff
@@ -110,3 +108,10 @@ def get(self):
                                     mapzen_api_key,
                                     self._username,
                                     self._orgname)
+
+    def _get_quota(self, mapzen_server_conf):
+        geocoding_quota = self._org_conf.get('geocoding_quota') or self._user_conf.get('geocoding_quota')
+        if not geocoding_quota:
+            geocoding_quota = mapzen_server_conf['geocoder']['monthly_quota']
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/refactor/service/mapzen_geocoder_config.py]: quick question: why don't we allow a user quota to be set to 0? (I saw a related test below)

Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/refactor/service/mapzen_geocoder_config.py]: We should let be 0 but avoid None. Good catch. Where is the related test?

Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/refactor/service/mapzen_geocoder_config.py]: Oh the acceptance test

**Comment by ethervoid** [comment]
Added:

- Check for different values of quota:
  - If it has a value or 0 it remains that way
  - If is '' or doesn't exists we left the quota to 0

- Added  more tests to cover all the providers in the config class to check for quota, service type and soft limit
- Change the Redis helper for tests to have all as string, like redis, instead of integers so we can check the conversion from string to int too

**Comment by rafatower** [comment]
LGTM :+1:
the conflict is just the version number, I can solve it. Other than that, time to think on acceptance criteria.

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py [hunk 2]
```diff
@@ -217,18 +214,16 @@ def __parse_config(self, filtered_config, db_config):
             self._isolines_provider = self.DEFAULT_PROVIDER
         self._geocoder_provider = filtered_config[self.GEOCODER_PROVIDER_KEY].lower()
         self._period_end_date = date_parse(filtered_config[self.PERIOD_END_DATE])
+        self._isolines_quota = self._get_effective_monthly_quota(self.QUOTA_KEY)
+        if filtered_config[self.SOFT_LIMIT_KEY].lower() == 'true':
+            self._soft_isolines_limit = True
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: glad this is "standardized" :)

**Comment by rafatower** [comment]
We have a "semantic conflict" between this and @jgoizueta 's [PR](https://github.com/CartoDB/dataservices-api/pull/329) that we have to address:
```
17:15:03 ======================================================================
17:15:03 ERROR: test_org_quota_quota_for_a_month_shorter_than_end_day (test.test_user_service.TestUserService)
17:15:03 ----------------------------------------------------------------------
17:15:03 Traceback (most recent call last):
17:15:03   File "/home/ubuntu/workspace/CartoDB-DATASERVICES-API-PR-testing/server/lib/python/cartodb_services/test/test_user_service.py", line 38, in test_org_quota_quota_for_a_month_shorter_than_end_day
17:15:03     us = self.__build_user_service('test_user', orgname='test_org', end_date=date(2016,1,31))
17:15:03   File "/home/ubuntu/workspace/CartoDB-DATASERVICES-API-PR-testing/server/lib/python/cartodb_services/test/test_user_service.py", line 149, in __build_user_service
17:15:03     end_date=end_date)
17:15:03   File "/home/ubuntu/workspace/CartoDB-DATASERVICES-API-PR-testing/server/lib/python/cartodb_services/test/test_helper.py", line 15, in build_redis_user_config
17:15:03     end_date_tz = end_date.replace(tzinfo=tzlocal())
17:15:03 TypeError: 'tzinfo' is an invalid keyword argument for this function
17:15:03 
17:15:03 ======================================================================
17:15:03 ERROR: test_user_quota_for_a_month_shorter_than_end_day (test.test_user_service.TestUserService)
17:15:03 ----------------------------------------------------------------------
17:15:03 Traceback (most recent call last):
17:15:03   File "/home/ubuntu/workspace/CartoDB-DATASERVICES-API-PR-testing/server/lib/python/cartodb_services/test/test_user_service.py", line 27, in test_user_quota_for_a_month_shorter_than_end_day
17:15:03     us = self.__build_user_service('test_user', end_date=date(2016,1,31))
17:15:03   File "/home/ubuntu/workspace/CartoDB-DATASERVICES-API-PR-testing/server/lib/python/cartodb_services/test/test_user_service.py", line 149, in __build_user_service
17:15:03     end_date=end_date)
17:15:03   File "/home/ubuntu/workspace/CartoDB-DATASERVICES-API-PR-testing/server/lib/python/cartodb_services/test/test_helper.py", line 15, in build_redis_user_config
17:15:03     end_date_tz = end_date.replace(tzinfo=tzlocal())
17:15:03 TypeError: 'tzinfo' is an invalid keyword argument for this function
17:15:03 
17:15:03 ----------------------------------------------------------------------
```

File: server/lib/python/cartodb_services/test/test_helper.py [hunk 3]
```diff
@@ -8,46 +9,60 @@
...
 
 
-def build_redis_user_config(redis_conn, username, quota=100, soft_limit=False,
-                            service="heremaps", isolines_quota=0,
-                            do_quota=None, soft_do_limit=None,
-                            do_general_quota=None, soft_do_general_limit=None,
+def build_redis_user_config(redis_conn, username, service, quota=100,
+                            soft_limit=False, provider="heremaps",
                             end_date=datetime.today()):
+    end_date_tz = end_date.replace(tzinfo=tzlocal())
```
Comment (rafatower) [inline server/lib/python/cartodb_services/test/test_helper.py]: question: why do we need this if all the internal timestamps are meant to be UTC?

**Comment by jgoizueta** [comment]
It seems `build_redis_org_config ` is now expecting a `datetime` (previously a `date` was also OK) and the tests I wrote provide a `date`

File: server/lib/python/cartodb_services/cartodb_services/metrics/config.py (hunk 2 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/cartodb_services/metrics/config.py]: Yep, at last.

File: server/lib/python/cartodb_services/cartodb_services/metrics/user.py [hunk 4]
```diff
@@ -2,21 +2,25 @@
...
     return date(year, month, valid_day)
 
+
 class UserMetricsService:
     """ Class to manage all the user info """
 
     SERVICE_GEOCODER_NOKIA = 'geocoder_here'
     SERVICE_GEOCODER_CACHE = 'geocoder_cache'
     SERVICE_HERE_ISOLINES = 'here_isolines'
+    SERVICE_MAPZEN_ISOLINES = 'mapzen_isolines'
```
Comment (jgoizueta) [inline server/lib/python/cartodb_services/cartodb_services/metrics/user.py]: I thought we had an isolines service (with the regrettable name of `here_isolines`) and two providers for it (here & mapzen). 😕 I'm sure @ethervoid knows what he's doing here, but @rafatower, can we double check it before deploying in his absence?

File: server/lib/python/cartodb_services/cartodb_services/mapzen/qps.py [hunk 5]
```diff
@@ -55,8 +55,8 @@ def call(self, fn, *args, **kwargs):
 
     def retry(self, first_request_time, retry_count):
         elapsed = datetime.now() - first_request_time
-        elapsed_miliseconds = (elapsed.total_seconds() * 1000.0)
-        if elapsed_miliseconds  > (self._retry_timeout * 1000.0):
+        elapsed_milliseconds = (elapsed.total_seconds() * 1000.0)
+        if elapsed_milliseconds  > (self._retry_timeout * 1000.0):
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/mapzen/qps.py]: how about not multiplying any of the quantities by `1000.0` and renaming it to `elapsed_seconds` ?

Comment (jgoizueta) [inline server/lib/python/cartodb_services/cartodb_services/mapzen/qps.py]: Totally, I see no reason to compute milliseconds. Funny how I was blinded by a silly typo (miliseconds) and didn't see that obvious redundant math.

**Pull request merged**